processing/src/main/java/io/druid/query/topn/TopNQueryEngine.java

```diff
@@ -112,8 +113,12 @@ public TopNQueryEngine(StupidPool<ByteBuffer> bufferPool)
       // A special TimeExtractionTopNAlgorithm is required, since DimExtractionTopNAlgorithm
       // currently relies on the dimension cardinality to support lexicographic sorting
       topNAlgorithm = new TimeExtractionTopNAlgorithm(capabilities, query);
-    } else if(selector.isHasExtractionFn()) {
-      topNAlgorithm = new DimExtractionTopNAlgorithm(capabilities, query);
+    } else if (selector.isHasExtractionFn()) {
+      if (canApplyExtractionInPost(query)) {
```

Comment:
So, I interpret that as you have now uncovered a bug that we weren't aware of and adding this code is adding a bunch of cruft to work around the bug instead of fix the bug ;)